### PR TITLE
Added `--mzxml-dir` option to load_study and load_msruns

### LINKED ISSUE #893: Add an mzXML directory option
### Inspiration

It is laborious to account for all subdirectories where mzXML files live, even if supplying a glob.

### Description

It would be nice to be able to supply an mzXML root directory, and walk the directory to find all the mzXML files.  Add an mzXML directory argument/option to:
- `StudyLoader`
- `load_study`
- `MSRunsLoader`
- `load_msruns`

### Alternatives

Provide a glob utilizing curly braces.

### Dependencies

None

### Comment

None

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### LINKED ISSUE #1086: Make `load_study` accept `mzxml` files (in addition to the study doc)
### Inspiration

See #1034

### Description

Make `load_study` accept `mzxml` files (in addition to the study doc)

### Alternatives

None

### Dependencies

None

### Comment

None

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

See affected issue(s).

## Affected Issues/Pull Requests

- Resolves #893
- Resolves #1086
- Merges into PR #1239

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/loaders/study_loader.py [hunk 1]
```diff
@@ -347,7 +352,18 @@ def __init__(self, *args, **kwargs):
...
+        Args:
+            None
+        Exceptions:
+            Raises:
+                ProgrammingError
+            Buffers:
+                None
+        Returns:
+            None
+        """
```
Comment (hepcat72) [inline DataRepo/loaders/study_loader.py:366]: C. Fleshed the docstring.

File: DataRepo/loaders/study_loader.py [hunk 2]
```diff
@@ -1015,6 +1031,21 @@ def get_supported_versions(cls) -> List[str]:
...
+        Exceptions:
+            Raises:
+                InvalidStudyDocVersion
+                UnknownStudyDocVersion
+                MultipleStudyDocVersions
+            Buffers:
+                None
+        Returns:
+            loader_class (StudyLoader): A derived class of StudyLoader matching the study doc version
+        """
```
Comment (hepcat72) [inline DataRepo/loaders/study_loader.py:1048]: C. Added docstring.

File: DataRepo/tests/loaders/test_msruns_loader.py [hunk 3]
```diff
@@ -1559,3 +1559,32 @@ def test_get_loaded_msrun_sample_dict(self):
         }
 
         self.assertDictEqual(expected, msrsd)
+
+    def test_get_mzxml_files_none(self):
+        mzxml_files = MSRunsLoader.get_mzxml_files()
+        self.assertEqual([], mzxml_files)
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_msruns_loader.py:1565]: C. Tests that when no args are supplied, it returns an empty list.

File: DataRepo/tests/loaders/test_msruns_loader.py [hunk 4]
```diff
@@ -1559,3 +1559,32 @@ def test_get_loaded_msrun_sample_dict(self):
...
+
+    def test_get_mzxml_files_files(self):
+        files = [
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/Br-xz971.mzXML",
+        ]
+        mzxml_files = MSRunsLoader.get_mzxml_files(
+            files=files,
+            dir="DataRepo/data/tests/small_obob",
+        )
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_msruns_loader.py:1575]: C. This tests that the `files` argument overrides the `dir` argument (`DataRepo/data/tests/small_obob` contains more than the 2 files passed via the `files` argument).

File: DataRepo/tests/loaders/test_study_loader.py [hunk 5]
```diff
@@ -270,6 +270,22 @@ def test_determine_matching_versions_v3(self):
...
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_lactate_neg_mzxmls/Br-xz971_neg.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_lactate_neg_mzxmls/BAT-xz971_neg.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/Br-xz971.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_lactate_mzxmls/Br-xz971.mzXML",
+            "DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_lactate_mzxmls/BAT-xz971.mzXML",
+        ]
+        self.assertEqual(
+            set(expected), set(sl.CustomLoaderKwargs.HEADERS["mzxml_files"])
+        )
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_study_loader.py:287]: C. This tests that `StudyLoader` passes the directory's `mzXML` files.

**Review by jcmatese — APPROVED** [review]
Seems good to me [new mzxml dir option, mutually exclusive with prior files argument, matching mzxml suffixes case-insensitively, mzxml arguments are both optional, added test and doc updates, etc.].

**Pull request merged**